pkg/cloudprovider/providers/azure/azure_managedDiskController.go

```diff
@@ -0,0 +1,366 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+	kwait "k8s.io/apimachinery/pkg/util/wait"
+)
+
+//ManagedDiskController : managed disk controller struct
+type ManagedDiskController struct {
+	common *controllerCommon
+}
+
+func newManagedDiskController(common *controllerCommon) (*ManagedDiskController, error) {
+	return &ManagedDiskController{common: common}, nil
+}
+
+//AttachManagedDisk : attach managed disk
+func (c *ManagedDiskController) AttachManagedDisk(nodeName string, diskURI string, cacheMode string) (int, error) {
+	// We don't need to validate if the disk is already attached
+	// to a different VM. The VM update call below will fail if
+	// it was attached somewhere else
+
+	// this behaviour is expected in (i.e during a k8s drain node call)
+	// k8s will evantually call detach->oldnode followed by attach->newnode
+	var vmData interface{}
+	vm, err := c.common.getArmVM(nodeName)
+
+	if err != nil {
+		return -1, err
+	}
+
+	if err := json.Unmarshal(vm, &vmData); err != nil {
+		return -1, err
+	}
+
+	fragment, ok := vmData.(map[string]interface{})
+	if !ok {
+		return -1, fmt.Errorf("convert vmData to map error")
+	}
+	// remove "resources" as ARM does not support PUT with "resources"
+	delete(fragment, "resources")
+
+	dataDisks, storageProfile, hardwareProfile, err := ExtractVMData(fragment)
+	if err != nil {
+		return -1, err
+	}
+	vmSize := hardwareProfile["vmSize"].(string)
+
+	managedVM := c.common.isManagedArmVM(storageProfile)
+	if !managedVM {
+		return -1, fmt.Errorf("azureDisk - error: attempt to attach managed disk %s to an unmanaged node  %s ", diskURI, nodeName)
+	}
+
+	// lock for findEmptyLun and append disk
+	var mutex = &sync.Mutex{}
```

Comment:
Can you help me understand why we need this lock? i understand we are trying to prevent race conditions on getting free-lun-number that may result in getting the same lun. 

The following applies in context:
1. on the slim chance that two+ attach operations got the same lun, arm will apply one and reject the rest which will be retried by the controller. 
2. The current lock slows down the entire attach operations and unnecessary/partially serialize it in sequence.
3. The lock is in fact does protect against anything, two attach op can get the same when op A: lock, got lun#1, unlock, then op A's go routine de-schedule before running actual attach-call for op b: which will go through lock, get lun#1 (not yet reserved by op A), unlocked.